content/blog/15-making-rust-attractive-for-writing-gtk-applications.md

```diff
@@ -0,0 +1,301 @@
+
+## Portals
+
+Modern Linux applications should make use of the [portals](https://flatpak.github.io/xdg-desktop-portal/) to request runtime permissions to access resources such as the camera, starting a screencast session or picking a file. 
+
+[ASHPD](https://crates.io/crates/ashpd) is nowadays the way to go for using portals from Rust as it provides a convenient & idomatic API on top of the DBus one.
+
+Code example from `ashpd-0.4.0-alpha.1` for using the color picker in a desktop environment agnostic way.
+```rust
+use ashpd::desktop::screenshot::ColorResponse;
+
+async fn run() -> ashpd::Result<()> {
+    let color = ColorResponse::builder().build().await?;
+    println!("({}, {}, {})", color.red(), color.green(), color.blue());
+    Ok(())
+}
+```
+
+
+## Keyring & oo7
+
+Applications that needs to store sensitive information such as passwords usually go through the [Secret service](https://specifications.freedesktop.org/secret-service/latest/) which is a protocol supported by both gnome-keyring and kwallet nowadays. There were multiple attempts to provide a Rust wrapper of the DBus API but sadly those were lacking either async support, they didn't provide integration with the secret portal or had no way to allow applications to migrate their secrets from the host keyring to the application sandboxed one.
+
+Those were the primary reasons we started working on [oo7](https://crates.io/crates/oo7), it is still on alpha stage, but should cover most of the use cases already.
+
+## GStreamer
+
+As part of the experiments I did when porting [Authenticator](https://gitlab.gnome.org/World/Authenticator) to GTK 4 was figuring out how can I replicate the internal GStreamer sink that GTK has to convert a video frame into a `gdk::MemoryTexture` in order to render it in some widget for QR codes scanning purposes.
+
+Jordan Petridis took over my WIP work and turned it into a proper GStreamer plugin written in Rust, see <https://gitlab.freedesktop.org/gstreamer/gst-plugins-rs/-/tree/main/video/gtk4> for an example on how to use it in your application.
+
+## Missing integrations
+
+### Better gettext support
+
+Currently, [gettext](https://www.gnu.org/software/gettext/) doesn't support Rust officially and most importantly, it doesn't like the `!` character, so we can't use string formatting for translatable texts. 
+
+Kévin Commaille have submitted a [patch](https://savannah.gnu.org/bugs/?56774) for upstream gettext but sadly it hasn't been reviewed yet :( For now people are working around this by replacing the variables names with <https://doc.rust-lang.org/std/primitive.str.html#method.replace>, not ideal but it is what we have for now.
+
+### Reduced boilerplate in subclassing code
+
+As mentioned above, subclassing code is still too verbose in some cases. Ideally, we would simplify most of it and it is probably one of the most confusing things you have to deal with as a beginner to the gtk-rs ecosystem.
```

Comment:
```suggestion
As mentioned above, subclassing code is still too verbose in some cases. Ideally, we would simplify most of it, since it is probably one of the most confusing things you have to deal with as a beginner to the gtk-rs ecosystem.
```